module/VuFind/src/VuFind/Captcha/Altcha.php

```diff
@@ -0,0 +1,119 @@
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  CAPTCHA
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Captcha;
+
+use AltchaOrg\Altcha\Altcha as AltchaController;
+use AltchaOrg\Altcha\ChallengeOptions;
+use AltchaOrg\Altcha\Hasher\Algorithm;
+use Laminas\Mvc\Controller\Plugin\Params;
+
+use function intval;
+
+/**
+ * Altcha proof-of-work CAPTCHA.
+ *
+ * @category VuFind
+ * @package  CAPTCHA
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class Altcha extends AbstractBase
+{
+    /**
+     * Constructor
+     *
+     * @param string                  $hmacKey Required HMAC key for challenge calculation and solution verification.
+     *
+     * @param Algorithm               $algorithm  Hashing algorithm to use (`SHA-1`, `SHA-256`, `SHA-512`, default: `SHA-256`).
+     * @param int                     $maxNumber  Maximum number for the random number generator (default: 1,000,000)
+     * @param null|\DateTimeInterface $expires    Optional expiration time for the challenge.
+     * @param ChallengeParams         $params     Optional URL-encoded query parameters.
+     * @param int<1, max>             $saltLength Length of the random salt (default: 12 bytes).
+     */
+    public function __construct(
+        protected string $hmacKey,
+        // Options for creation of a new challenge
+        protected Algorithm $algorithm = Algorithm::SHA256,
+        protected int $maxNumber = AltchaController::DEFAULT_MAX_NUMBER,
+        protected ?\DateTimeInterface $expires = null,
+        protected array $params = [],
+        protected int $saltLength = AltchaController::DEFAULT_SALT_LENGTH,
+    ) {
+    }
+
+    /**
+     * Get list of URLs with JS dependencies to load for the active CAPTCHA type.
+     *
+     * @return array
+     */
+    public function getJsIncludes(): array
+    {
+        return ['vendor/altcha.js', 'vendor/altcha-i18n.js'];
+    }
+
+    /**
+     * Generate challenge
+     *
+     * @return Challenge
+     */
+    public function getChallenge()
+    {
+        $altcha = new AltchaController($this->hmacKey);
```

Comment:
Do you need to create a new Altcha object for every challenge? If not, maybe it would be better to build this object in the factory and inject it to the constructor. This would make testing easier, since it would allow us to mock the object; might also have customization-related benefits.